org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/IWorkspaceConfigFactory.xtend

```diff
@@ -30,16 +33,92 @@ interface IWorkspaceConfigFactory {
  * @since 2.11
  */
 class ProjectWorkspaceConfigFactory implements IWorkspaceConfigFactory {
-    
-    override getWorkspaceConfig(URI workspaceBaseURI) {
-    	if (workspaceBaseURI === null) {
-			val projectConfig = new InMemoryProjectConfig
-			return projectConfig.workspaceConfig
-    	} else {
-	        val projectConfig = new FileProjectConfig(workspaceBaseURI)
-	        projectConfig.sourceFolders += new FileSourceFolder(projectConfig, '.')
-	        return projectConfig.workspaceConfig
-        }
-    }
-    
+
+	override getWorkspaceConfig(URI workspaceBaseURI) {
+		val workspaceConfig = new MultiProjectWorkspaceConfig
+		workspaceConfig.findProjects(workspaceBaseURI)
+		return workspaceConfig
+	}
+
+	def void findProjects(MultiProjectWorkspaceConfig workspaceConfig, URI uri) {
+		workspaceConfig.addProject(uri)
+	}
+
+}
+
+class MultiProjectWorkspaceConfig implements IWorkspaceConfig {
+
+	val unknown = new UnknownProjectConfig(this)
+	val projects = <String, IProjectConfig>newHashMap
+
+	new() {
+		addProject(unknown)
+	}
+
+	override getProjects() {
+		projects.values.toSet
+	}
+
+	def FileProjectConfig addProject(URI uri, String name) {
```

Comment:
These factory methods should go to `ProjectWorkspaceConfigFactory` so it can be overridden through DI.
Also, this is not a heavily used API so we don't need all the convenient overrides.